rpc_util.go

```diff
@@ -231,6 +239,48 @@ func PerRPCCredentials(creds credentials.PerRPCCredentials) CallOption {
 	})
 }
 
+// CallContentSubtype returns a CallOption that will set the content-subtype
+// for a call. For example, if content-subtype is "json", the Content-Type
+// over the wire will be "application/grpc+json". See Content-Type on
+// https://grpc.io/docs/guides/wire.html#requests for more details. Note that
+// content-subtypes are case-insensitive.
+//
+// If CallCustomCodec is not also used, the content-subtype will be used to
+// look up the Codec to use in the registry controlled by RegisterCodec. See
+// the documention on RegisterCodec for details on registration. If no such
+// Codec is found, the call will result in an error with code codes.Internal.
+//
+// If CallCustomCodec is used, this Codec will be used for all request and
+// response messages, with the content-subtype still being set to the given
+// content-subtype for all requests.
+func CallContentSubtype(contentSubtype string) CallOption {
+	contentSubtype = strings.ToLower(contentSubtype)
+	return beforeCall(func(c *callInfo) error {
+		if contentSubtype == "" {
+			return errEmptyContentSubtypeCallContentSubtype
+		}
+		c.contentSubtype = contentSubtype
+		return nil
+	})
+}
+
+// CallCustomCodec returns a CallOption that will set the given Codec to be
+// used for all request and response messages for a call. The result of calling
+// String() will be used as the content-subtype in a case-insensitive manner.
+// See Content-Type on https://grpc.io/docs/guides/wire.html#requests for more
+// details. Also see the documentation on RegisterCodec and CallContentSubtype
+// for more details on the interaction between Codec and content-subtype.
+func CallCustomCodec(codec Codec) CallOption {
+	contentSubtype := codec.String()
+	return beforeCall(func(c *callInfo) error {
+		if contentSubtype == "" {
```

Comment:
Similar to above.  It's probably better to continue without setting the content subtype than error out.  If there was some way to error `CallCustomCodec()` itself, that would make sense, but given that checking inside `beforeCall` delays the check, I think we should try to continue if possible.

Reply:
Let's chat offline perhaps? I feel moderately strongly we should be erroring if we use an empty string to this option or the other option.

Reply:
I think we need to avoid an error here to prevent a backward-compatibility breakage.  I.e. we _currently_ ignore the `String()` of the codec provided entirely.  Now that I think about it more, we also really shouldn't set the content sub-type at all when this setting is used, either.  (Note that this is how `WithCodec` will be implemented.)  There's a reasonable possibility of breaking anyone using custom codecs and non-Go peers if we change it.

We need to provide this for backward compatibility and also for the use case you're interested in, but we should discourage users from using it, and direct them strongly toward the `CallContentSubtype()` + `encoding.RegisterCodec()` APIs instead.

If you disagree with the above and want to debate further, I'm happy to, but yes, let's take it offline.